In [ ]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
import pandas as pd
import re
from collections import Counter
import numpy as np


In [ ]:
# Path to your CSV file in Google Drive
file_path = '/content/drive/My Drive/CloudComputing/NationalLibrarySingaporeDigitisedBooksandMagazines.csv'

# Read CSV into dataframe
df = pd.read_csv(file_path)

<ipython-input-3-0f3c981130d5>:5: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


In [ ]:


# Set to store unique values
unique_values = set()
values_list = list()

# Iterate through unique values
for value in df['subject_singheritage']:
    if isinstance(value, float):
        continue
    else:
        # Remove {<intvalue>} if present in the string
        modified_value = re.sub(r'\{\d+\}', '', value)  # Remove {<intvalue>}
        modified_value = re.sub(r'\|', '', modified_value)  # Remove {<intvalue>}
        modified_value = re.sub(r'[()]', '', modified_value).strip().lower()  # Remove parentheses, strip whitespace, and convert to lowercase
        unique_values.add(modified_value)
        values_list.append(modified_value)


for value in sorted(unique_values):
    print(value)


print()
print()
print()
print()


# Define common words to ignore
common_words_to_ignore = {"and", "of", "the", "is", "a", "this", "that", "to"}  # Add more common words as needed

# Split each string into words and flatten them into a single list, while filtering out common words
words = [word for string in values_list for word in string.split() if word.lower() not in common_words_to_ignore]

# Count occurrences of each word
word_counts = Counter(words)

# Sort words based on occurrence counts in descending order
sorted_words = sorted(word_counts, key=lambda x: word_counts[x], reverse=True)

# Print common words in descending order of occurrence
print("Common words in descending order of occurrence:")
for word in sorted_words:
    print(word, ":", word_counts[word])

1819-1826 founding and early years 1826-1867 straits settlements
1819-1826 founding and early years 1826-1867 straits settlements 1867-1914 port city 1914-1942 cosmopolitan city pioneers heritage and culture
1819-1826 founding and early years 1826-1867 straits settlements national security
1819-1826 founding and early years colonial administrators politics and government
1819-1826 founding and early years law
1819-1826 founding and early years pre-1819 temasek
1826-1867 straits settlements
1826-1867 straits settlements 1867-1914 port city christianity heritage and culture personalities
1826-1867 straits settlements colonial administrators indochina
1826-1867 straits settlements public administration trade and industry colonial administrators organisations business enterprises
1867-1914 port city
1867-1914 port city 1914-1942 cosmopolitan city malay archipelago
1867-1914 port city 1914-1942 cosmopolitan city organisations
1867-1914 port city description and travel malay archipelago
1867

In [ ]:
# Define the set of words to search for
words_to_search = {'Education', 'school', 'schools', 'arts', 'art', 'science', 'technology', 'conferences', 'conventions', 'children',
                   'heritage', 'culture', 'primary', 'secondary', 'textbooks', 'textbook', 'history', 'sports', 'ict', 'libraries',
                   'literature', 'dialects', 'peotry', 'universities', 'university', 'college', 'colleges', 'youth' }

# Create a new dataframe to store rows where the specific word is present
new_df = pd.DataFrame(columns=df.columns)

# Iterate through the rows of the original DataFrame
for index, row in df.iterrows():
    # Check if any word from the list is present in both columns
    if isinstance(row['subject_lcsh'], float) or isinstance(row['subject_singheritage'], float):
        continue
    elif any(word in row['subject_lcsh'].lower() or word in row['subject_singheritage'].lower() for word in words_to_search):
        new_df = new_df.append(row, ignore_index=True)

<ipython-input-5-25ef6e3824d3>:15: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  new_df = new_df.append(row, ignore_index=True)


In [ ]:
# Saving DataFrame to CSV
new_df.to_csv('/content/drive/My Drive/CloudComputing/newdatatemp.csv', index=False)

In [ ]:
file_path_new = '/content/drive/My Drive/CloudComputing/newdatatemp.csv'

# Read CSV into dataframe
newdata_df = pd.read_csv(file_path_new)

In [ ]:
display(newdata_df)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri identifier_isbn  \
0      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
1      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
2      https://eresources.nlb.gov.sg/main/malay-ebook...             NaN   
3      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
4      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
...                                                  ...             ...   
13712  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13713  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13714  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13715  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13716  https://eresources.nlb.gov.sg/printheritage/de...             NaN   

      identifier_issn                                              title  \
0                 NaN                                  原产报, 66 (一九七〇年一月)   
1                 NaN                Meridian Junior College album, 2014   
2                 NaN                             Menjelajah zaman silam   
3                 NaN                                          கவித்தொகை   
4                 NaN           Raffles Girls' Primary School, Apr. 2008   
...               ...                                                ...   
13712             NaN  Fregatten Eugenies resa omkring jorden, åren ...   
13713             NaN                     Maris Stella High School, 1999   
13714             NaN                  Monk's Hill School magazine, 1956   
13715             NaN        Lianhua Primary School annual magazine 2009   
13716             NaN                                     The blossoming   

                                       alternative_title  \
0      Primary production bulletin|Berita hasil kampo...   
1                                        Meridian family   
2                                                    NaN   
3                                             Kavittokai   
4                                                    NaN   
...                                                  ...   
13712                                                NaN   
13713                            海星中学|Hai xing zhong xue   
13714                                                NaN   
13715                                                NaN   
13716                                                NaN   

                    creator creator_people creator_people_cn  \
0                       NaN            NaN               NaN   
1                       NaN            NaN               NaN   
2         Krishnan Panikker            NaN               NaN   
3      அன்பழகன், மா., 1943-            NaN               NaN   
4                       NaN            NaN               NaN   
...                     ...            ...               ...   
13712                   NaN            NaN               NaN   
13713                   NaN            NaN               NaN   
13714    Monk's Hill School            NaN               NaN   
13715                   NaN            NaN               NaN   
13716                   NaN            NaN               NaN   

                                   creator_orgs  ...  \
0                                           NaN  ...   
1            Meridian Junior Colleg

In [ ]:
# # Columns to fill NaN values
# columns_to_fillna = ['creator', 'creator_people', 'creator_people_cn', 'creator_orgs']

# # Fill NaN values with empty strings for specified columns
# newdata_df[columns_to_fillna] = newdata_df[columns_to_fillna].fillna('')

# Define a function to merge non-empty strings with comma
def merge_with_comma(x, y):
    if pd.notna(x) and pd.notna(y):
        return str(x) + ', ' + str(y)
    elif pd.notna(x):
        return x
    elif pd.notna(y):
        return y
    else:
        return np.nan

# Merge text of four columns into a new column
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['creator'], row['creator_people']), axis=1)
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_creator'], row['creator_people_cn']), axis=1)
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_creator'], row['creator_orgs']), axis=1)


display(newdata_df)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri identifier_isbn  \
0      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
1      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
2      https://eresources.nlb.gov.sg/main/malay-ebook...             NaN   
3      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
4      https://eresources.nlb.gov.sg/printheritage/de...             NaN   
...                                                  ...             ...   
13712  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13713  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13714  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13715  https://eresources.nlb.gov.sg/printheritage/de...             NaN   
13716  https://eresources.nlb.gov.sg/printheritage/de...             NaN   

      identifier_issn                                              title  \
0                 NaN                                  原产报, 66 (一九七〇年一月)   
1                 NaN                Meridian Junior College album, 2014   
2                 NaN                             Menjelajah zaman silam   
3                 NaN                                          கவித்தொகை   
4                 NaN           Raffles Girls' Primary School, Apr. 2008   
...               ...                                                ...   
13712             NaN  Fregatten Eugenies resa omkring jorden, åren ...   
13713             NaN                     Maris Stella High School, 1999   
13714             NaN                  Monk's Hill School magazine, 1956   
13715             NaN        Lianhua Primary School annual magazine 2009   
13716             NaN                                     The blossoming   

                                       alternative_title  \
0      Primary production bulletin|Berita hasil kampo...   
1                                        Meridian family   
2                                                    NaN   
3                                             Kavittokai   
4                                                    NaN   
...                                                  ...   
13712                                                NaN   
13713                            海星中学|Hai xing zhong xue   
13714                                                NaN   
13715                                                NaN   
13716                                                NaN   

                    creator creator_people creator_people_cn  \
0                       NaN            NaN               NaN   
1                       NaN            NaN               NaN   
2         Krishnan Panikker            NaN               NaN   
3      அன்பழகன், மா., 1943-            NaN               NaN   
4                       NaN            NaN               NaN   
...                     ...            ...               ...   
13712                   NaN            NaN               NaN   
13713                   NaN            NaN               NaN   
13714    Monk's Hill School            NaN               NaN   
13715                   NaN            NaN               NaN   
13716                   NaN            NaN               NaN   

                                   creator_orgs  ...  \
0                                           NaN  ...   
1            Meridian Junior Colleg

In [ ]:
newdata_df['identifier_uri'] = 'https://www.nlb.gov.sg/main/book-detail?cmsuuid=' + df['identifier_uuid'].astype(str)
display(newdata_df)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri identifier_isbn  \
0      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
1      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
2      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
3      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
4      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
...                                                  ...             ...   
13712  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13713  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13714  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13715  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13716  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   

      identifier_issn                                              title  \
0                 NaN                                  原产报, 66 (一九七〇年一月)   
1                 NaN                Meridian Junior College album, 2014   
2                 NaN                             Menjelajah zaman silam   
3                 NaN                                          கவித்தொகை   
4                 NaN           Raffles Girls' Primary School, Apr. 2008   
...               ...                                                ...   
13712             NaN  Fregatten Eugenies resa omkring jorden, åren ...   
13713             NaN                     Maris Stella High School, 1999   
13714             NaN                  Monk's Hill School magazine, 1956   
13715             NaN        Lianhua Primary School annual magazine 2009   
13716             NaN                                     The blossoming   

                                       alternative_title  \
0      Primary production bulletin|Berita hasil kampo...   
1                                        Meridian family   
2                                                    NaN   
3                                             Kavittokai   
4                                                    NaN   
...                                                  ...   
13712                                                NaN   
13713                            海星中学|Hai xing zhong xue   
13714                                                NaN   
13715                                                NaN   
13716                                                NaN   

                    creator creator_people creator_people_cn  \
0                       NaN            NaN               NaN   
1                       NaN            NaN               NaN   
2         Krishnan Panikker            NaN               NaN   
3      அன்பழகன், மா., 1943-            NaN               NaN   
4                       NaN            NaN               NaN   
...                     ...            ...               ...   
13712                   NaN            NaN               NaN   
13713                   NaN            NaN               NaN   
13714    Monk's Hill School            NaN               NaN   
13715                   NaN            NaN               NaN   
13716                   NaN            NaN               NaN   

                                   creator_orgs  ...  \
0                                           NaN  ...   
1            Meridian Junior Colleg

Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.
Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.


In [ ]:
newdata_df.drop(columns=['identifier_image_uri'], inplace=True)
newdata_df['identifier_image_uri'] = 'https://eservice.nlb.gov.sg/bookcoverwrapper/cover/' + df['identifier_uuid'].astype(str) + '?s=LG&type=UUID'
display(newdata_df)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri identifier_isbn  \
0      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
1      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
2      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
3      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
4      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
...                                                  ...             ...   
13712  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13713  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13714  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13715  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13716  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   

      identifier_issn                                              title  \
0                 NaN                                  原产报, 66 (一九七〇年一月)   
1                 NaN                Meridian Junior College album, 2014   
2                 NaN                             Menjelajah zaman silam   
3                 NaN                                          கவித்தொகை   
4                 NaN           Raffles Girls' Primary School, Apr. 2008   
...               ...                                                ...   
13712             NaN  Fregatten Eugenies resa omkring jorden, åren ...   
13713             NaN                     Maris Stella High School, 1999   
13714             NaN                  Monk's Hill School magazine, 1956   
13715             NaN        Lianhua Primary School annual magazine 2009   
13716             NaN                                     The blossoming   

                                       alternative_title  \
0      Primary production bulletin|Berita hasil kampo...   
1                                        Meridian family   
2                                                    NaN   
3                                             Kavittokai   
4                                                    NaN   
...                                                  ...   
13712                                                NaN   
13713                            海星中学|Hai xing zhong xue   
13714                                                NaN   
13715                                                NaN   
13716                                                NaN   

                    creator creator_people creator_people_cn  \
0                       NaN            NaN               NaN   
1                       NaN            NaN               NaN   
2         Krishnan Panikker            NaN               NaN   
3      அன்பழகன், மா., 1943-            NaN               NaN   
4                       NaN            NaN               NaN   
...                     ...            ...               ...   
13712                   NaN            NaN               NaN   
13713                   NaN            NaN               NaN   
13714    Monk's Hill School            NaN               NaN   
13715                   NaN            NaN               NaN   
13716                   NaN            NaN               NaN   

                                   creator_orgs  ...  \
0                                           NaN  ...   
1            Meridian Junior Colleg

In [ ]:
# Saving DataFrame to CSV
# newdata_df.to_csv('/content/drive/My Drive/CloudComputing/NLB_updated.csv', index=False)
newdata_df = pd.read_csv('/content/drive/My Drive/CloudComputing/NLB_updated.csv')


In [ ]:
# Count of null values in each column
null_counts = newdata_df.isnull().sum()

# Print the counts
print("Count of null values in each column:")
print(null_counts)

Count of null values in each column:
identifier_uuid              0
identifier_uri               0
identifier_isbn          13665
identifier_issn          13701
title                        0
alternative_title         7513
creator                  11669
creator_people           12750
creator_people_cn        13684
creator_orgs              7866
creator_orgs_cn          13477
creator_lcna             13717
contributor               2587
contributor_people       13573
contributor_people_cn    13608
contributor_orgs         13132
contributor_orgs_cn      13262
contributor_lcna         13535
digital_publisher          166
original_publisher         847
is_referenced_by          5902
date_created               552
description               7768
abstract                    57
subject_lcsh                 0
subject_singheritage         0
language                     8
ispartof_collections       140
nlb_type                     0
rights                      15
access_rights             1516
me

identifier_isbn
identifier_issn
contributor_people       13573
contributor_people_cn    13608
contributor_orgs         13132
contributor_orgs_cn      13262
contributor_lcna         13535
creator                  11669
creator_people           12750
creator_people_cn        13684
creator_orgs              7866
creator_orgs_cn          13477
creator_lcna             13717

In [ ]:
display(newdata_df)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri identifier_isbn  \
0      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
1      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
2      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
3      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
4      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
...                                                  ...             ...   
13712  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13713  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13714  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13715  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13716  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   

      identifier_issn                                              title  \
0                 NaN                                  原产报, 66 (一九七〇年一月)   
1                 NaN                Meridian Junior College album, 2014   
2                 NaN                             Menjelajah zaman silam   
3                 NaN                                          கவித்தொகை   
4                 NaN           Raffles Girls' Primary School, Apr. 2008   
...               ...                                                ...   
13712             NaN  Fregatten Eugenies resa omkring jorden, åren ...   
13713             NaN                     Maris Stella High School, 1999   
13714             NaN                  Monk's Hill School magazine, 1956   
13715             NaN        Lianhua Primary School annual magazine 2009   
13716             NaN                                     The blossoming   

                                       alternative_title  \
0      Primary production bulletin|Berita hasil kampo...   
1                                        Meridian family   
2                                                    NaN   
3                                             Kavittokai   
4                                                    NaN   
...                                                  ...   
13712                                                NaN   
13713                            海星中学|Hai xing zhong xue   
13714                                                NaN   
13715                                                NaN   
13716                                                NaN   

                    creator creator_people creator_people_cn  \
0                       NaN            NaN               NaN   
1                       NaN            NaN               NaN   
2         Krishnan Panikker            NaN               NaN   
3      அன்பழகன், மா., 1943-            NaN               NaN   
4                       NaN            NaN               NaN   
...                     ...            ...               ...   
13712                   NaN            NaN               NaN   
13713                   NaN            NaN               NaN   
13714    Monk's Hill School            NaN               NaN   
13715                   NaN            NaN               NaN   
13716                   NaN            NaN               NaN   

                                   creator_orgs  ...  \
0                                           NaN  ...   
1            Meridian Junior Colleg

In [ ]:
def merge_with_comma(x, y):
    if pd.notna(x) and pd.notna(y):
        return str(x) + ', ' + str(y)
    elif pd.notna(x):
        return x
    elif pd.notna(y):
        return y
    else:
        return np.nan

# Merge text of columns into a new column

# Drop existing merged_creator and merged_contributor columns if they exist
if 'merged_creator' in newdata_df.columns:
    newdata_df.drop(columns=['merged_creator'], inplace=True)
if 'merged_contributor' in newdata_df.columns:
    newdata_df.drop(columns=['merged_contributor'], inplace=True)

newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['creator'], row['creator_people']), axis=1)
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_creator'], row['creator_people_cn']), axis=1)
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_creator'], row['creator_orgs']), axis=1)
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_creator'], row['creator_orgs_cn']), axis=1)
newdata_df['merged_creator'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_creator'], row['creator_lcna']), axis=1)

newdata_df['merged_contributor'] = newdata_df.apply(lambda row: merge_with_comma(row['contributor'], row['contributor_people']), axis=1)
newdata_df['merged_contributor'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_contributor'], row['contributor_people_cn']), axis=1)
newdata_df['merged_contributor'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_contributor'], row['contributor_orgs']), axis=1)
newdata_df['merged_contributor'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_contributor'], row['contributor_orgs_cn']), axis=1)
newdata_df['merged_contributor'] = newdata_df.apply(lambda row: merge_with_comma(row['merged_contributor'], row['contributor_lcna']), axis=1)


display(newdata_df)

# Count of null values in each column
null_counts = newdata_df.isnull().sum()

# Print the counts
print("Count of null values in each column:")
print(null_counts)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri identifier_isbn  \
0      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
1      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
2      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
3      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
4      https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
...                                                  ...             ...   
13712  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13713  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13714  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13715  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   
13716  https://www.nlb.gov.sg/main/book-detail?cmsuui...             NaN   

      identifier_issn                                              title  \
0                 NaN                                  原产报, 66 (一九七〇年一月)   
1                 NaN                Meridian Junior College album, 2014   
2                 NaN                             Menjelajah zaman silam   
3                 NaN                                          கவித்தொகை   
4                 NaN           Raffles Girls' Primary School, Apr. 2008   
...               ...                                                ...   
13712             NaN  Fregatten Eugenies resa omkring jorden, åren ...   
13713             NaN                     Maris Stella High School, 1999   
13714             NaN                  Monk's Hill School magazine, 1956   
13715             NaN        Lianhua Primary School annual magazine 2009   
13716             NaN                                     The blossoming   

                                       alternative_title  \
0      Primary production bulletin|Berita hasil kampo...   
1                                        Meridian family   
2                                                    NaN   
3                                             Kavittokai   
4                                                    NaN   
...                                                  ...   
13712                                                NaN   
13713                            海星中学|Hai xing zhong xue   
13714                                                NaN   
13715                                                NaN   
13716                                                NaN   

                    creator creator_people creator_people_cn  \
0                       NaN            NaN               NaN   
1                       NaN            NaN               NaN   
2         Krishnan Panikker            NaN               NaN   
3      அன்பழகன், மா., 1943-            NaN               NaN   
4                       NaN            NaN               NaN   
...                     ...            ...               ...   
13712                   NaN            NaN               NaN   
13713                   NaN            NaN               NaN   
13714    Monk's Hill School            NaN               NaN   
13715                   NaN            NaN               NaN   
13716                   NaN            NaN               NaN   

                                   creator_orgs  ...  \
0                                           NaN  ...   
1            Meridian Junior Colleg

Count of null values in each column:
identifier_uuid              0
identifier_uri               0
identifier_isbn          13665
identifier_issn          13701
title                        0
alternative_title         7513
creator                  11669
creator_people           12750
creator_people_cn        13684
creator_orgs              7866
creator_orgs_cn          13477
creator_lcna             13717
contributor               2587
contributor_people       13573
contributor_people_cn    13608
contributor_orgs         13132
contributor_orgs_cn      13262
contributor_lcna         13535
digital_publisher          166
original_publisher         847
is_referenced_by          5902
date_created               552
description               7768
abstract                    57
subject_lcsh                 0
subject_singheritage         0
language                     8
ispartof_collections       140
nlb_type                     0
rights                      15
access_rights             1516
id

In [ ]:
newdata_df.drop(columns=['identifier_isbn'], inplace=True)
newdata_df.drop(columns=['identifier_issn'], inplace=True)
newdata_df.drop(columns=['alternative_title'], inplace=True)
newdata_df.drop(columns=['creator'], inplace=True)
newdata_df.drop(columns=['creator_people'], inplace=True)
newdata_df.drop(columns=['creator_people_cn'], inplace=True)
newdata_df.drop(columns=['creator_orgs'], inplace=True)
newdata_df.drop(columns=['creator_orgs_cn'], inplace=True)
newdata_df.drop(columns=['creator_lcna'], inplace=True)
newdata_df.drop(columns=['contributor'], inplace=True)
newdata_df.drop(columns=['contributor_people'], inplace=True)
newdata_df.drop(columns=['contributor_people_cn'], inplace=True)
newdata_df.drop(columns=['contributor_orgs'], inplace=True)
newdata_df.drop(columns=['contributor_orgs_cn'], inplace=True)
newdata_df.drop(columns=['contributor_lcna'], inplace=True)
newdata_df.drop(columns=['is_referenced_by'], inplace=True)



In [ ]:
display(newdata_df)

identifier_uuid  \
0      00000c04-4b98-4c0f-935f-3e9227cd71bb   
1      000146e1-723f-474a-9afd-eee90c082ea9   
2      012c407b-d6f1-4c7b-82e8-7ade9710c0bc   
3      00097dfe-d472-4d1a-97fc-343e76143234   
4      000a40db-7c79-42e0-8130-9b17c4d8865a   
...                                     ...   
13712  ffe4bd3d-3554-4244-a519-2bf4f7c81626   
13713  ffe6b084-91c8-48d1-8b7e-67deefa75663   
13714  ffef9eb0-f06b-48af-8f98-e6a5c2266c93   
13715  ffefae58-1ce3-4ac9-a86e-203ae62457a3   
13716  fff7ce6b-90ef-499a-bc2d-309df406f0cc   

                                          identifier_uri  \
0      https://www.nlb.gov.sg/main/book-detail?cmsuui...   
1      https://www.nlb.gov.sg/main/book-detail?cmsuui...   
2      https://www.nlb.gov.sg/main/book-detail?cmsuui...   
3      https://www.nlb.gov.sg/main/book-detail?cmsuui...   
4      https://www.nlb.gov.sg/main/book-detail?cmsuui...   
...                                                  ...   
13712  https://www.nlb.gov.sg/main/book-detail?cmsuui...   
13713  https://www.nlb.gov.sg/main/book-detail?cmsuui...   
13714  https://www.nlb.gov.sg/main/book-detail?cmsuui...   
13715  https://www.nlb.gov.sg/main/book-detail?cmsuui...   
13716  https://www.nlb.gov.sg/main/book-detail?cmsuui...   

                                                   title  \
0                                      原产报, 66 (一九七〇年一月)   
1                    Meridian Junior College album, 2014   
2                                 Menjelajah zaman silam   
3                                              கவித்தொகை   
4               Raffles Girls' Primary School, Apr. 2008   
...                                                  ...   
13712  Fregatten Eugenies resa omkring jorden, åren ...   
13713                     Maris Stella High School, 1999   
13714                  Monk's Hill School magazine, 1956   
13715        Lianhua Primary School annual magazine 2009   
13716                                     The blossoming   

                      digital_publisher                 original_publisher  \
0      National Library Board Singapore  原产局|Primary Production Department   
1      National Library Board Singapore            Meridian Junior College   
2                             DB United                    Penerbit Idaman   
3      National Library Board Singapore             புதுமைத்தேனீ பதிப்பகம்   
4      National Library Board Singapore      Raffles Girls' Primary School   
...                                 ...                                ...   
13712  National Library Board Singapore                         A. Bonnier   
13713  National Library Board Singapore           Maris Stella High School   
13714  National Library Board Singapore                 Monk's Hill School   
13715  National Library Board Singapore             Lianhua Primary School   
13716  National Library Board Singapore            Beatty Secondary School   

      date_created                                        description  \
0          1970-01  Text in traditional Chinese script. Partial te...   
1             2014                    Title devised by Library staff.   
2             2003                                                NaN   
3             2012                          பாரதி அச்சகம், சென்னை 41.   
4          2008-04                                                NaN   
...            ...                                                ...   
13712  1854 / 1855                                                NaN   
13713         1999                           Partial text in Chinese.   
13714          NaN          Partial text in Malay, Chinese and Tamil.   
13715         2009                                                NaN   
13716         1978                                                NaN   

                                                abstract  \
0      《原产报》是新加坡原产局出版的月刊, 介绍农牧技术性知识, 指导农业科学化应用与经营知识. ...   
1      This contains year one and year two class phot...   
2      Pakcik Nutcracker ad

In [ ]:
# Count of null values in each column
null_counts = newdata_df.isnull().sum()

# Print the counts
print("Count of null values in each column:")
print(null_counts)

Count of null values in each column:
identifier_uuid            0
identifier_uri             0
title                      0
digital_publisher        166
original_publisher       847
date_created             552
description             7768
abstract                  57
subject_lcsh               0
subject_singheritage       0
language                   8
ispartof_collections     140
nlb_type                   0
rights                    15
access_rights           1516
identifier_image_uri       0
merged_creator          4881
merged_contributor      2587
dtype: int64


In [ ]:
newdata_df.to_csv('NLB_updated.csv', index=False)
